spec.html

```diff
@@ -1564,11 +1563,14 @@ <h1>Number::multiply ( _x_, _y_ )</h1>
               In the remaining cases, where neither an infinity nor *NaN* is involved, the product is computed and rounded to the nearest representable value using IEEE 754-2019 roundTiesToEven mode. If the magnitude is too large to represent, the result is then an infinity of appropriate sign. If the magnitude is too small to represent, the result is then a zero of appropriate sign. The ECMAScript language requires support of gradual underflow as defined by IEEE 754-2019.
             </li>
           </ul>
+          <emu-note>
+            <p>Finite-precision multiplication is commutative, but not always associative.</p>
+          </emu-note>
         </emu-clause>
 
         <emu-clause id="sec-numeric-types-number-divide" oldids="sec-applying-the-div-operator">
           <h1>Number::divide ( _x_, _y_ )</h1>
-          <p>The `/` |MultiplicativeOperator| performs division, producing the quotient of _x_ and _y_. _x_ is the dividend and _y_ is the divisor. ECMAScript does not perform integer division. The operands and result of all division operations are double-precision floating-point numbers. The result of division is determined by the specification of IEEE 754-2019 arithmetic:</p>
+          <p>The abstract operation Number::divide takes arguments _x_ (a Number) and _y_ (a Number). It performs division, producing the quotient of _x_ and _y_; _x_ is the dividend and _y_ is the divisor. The result is determined by the specification of IEEE 754-2019 arithmetic:</p>
```

Comment:
```suggestion
          <p>The abstract operation Number::divide takes arguments _x_ (a Number) and _y_ (a Number). It performs division, producing the quotient of _x_ and _y_; _x_ is the dividend and _y_ is the divisor. The result is determined by the specification of IEEE 754-2019 binary double-precision arithmetic:</p>
```

We should either use this specificity in all of these AOs or in none of them.

Reply:
Yeah, this is the kind of (lack of) consistency I was talking about in the commit message.

When you say "all of these AOs", I'm guessing you don't mean every Number::foo method, but just the ones that delegate (at least some) semantics to IEEE 754: multiply, divide, remainder, add.

The general form for these is:
```
The result of <operation> is (governed|determined) (by|using) <spec>:
```
where `<operation>` is:
- `a floating-point multiplication`
- `division`
- `an ECMAScript floating-point remainder operation`
- `an addition`

and `<spec>` is:
- `the rules of IEEE 754-2019 binary double-precision arithmetic`
- `the specification of IEEE 754-2019 arithmetic`
- `the rules of IEEE arithmetic`
- `the rules of IEEE 754-2019 binary double-precision arithmetic`

Let me know if there's a specific form that the editors would like.

Reply:
> When you say "all of these AOs", I'm guessing you don't mean every Number::foo method

Correct.

> Let me know if there's a specific form that the editors would like.

I'm not even really sure we need this, since the algorithm makes it clear at the point when it defers to IEEE 754. But if we did include something, my preference is something like "The result of <operation> is determined in part by the rules of binary double-precision arithmetic in IEEE 754-2019" or "The result of <operation> is determined by the rules of binary double-precision arithmetic in IEEE 754-2019 for finite, non-zero inputs".

Reply:
> I'm not even really sure we need this, since the algorithm makes it clear at the point when it defers to IEEE 754.

I wouldn't say it's really clear. E.g., in Number::multiply, the final bullet says "In the remaining cases, ..., the product is computed and rounded to the nearest representable value using IEEE 754-2019 roundTiesToEven mode." So yes, there's an explicit reference to IEEE 754, but only for the rounding. Other than that, it's just "the product is computed".

Now, if you wanted to move the "as determined by IEEE 754" formula from the preamble down to the final bullet (and have it explicitly apply to the computation of the product/sum/whatever), that'd be fine too.

Reply:
Yeah then I'd say we should make the necessary clarification in the algorithm steps. The algorithm steps should be clear on their own without relying on cues from the preamble.

Reply:
@michaelficarra Do you want to open an issue (or PR) about clarifying the algorithm steps? Doesn't seem to me that it belongs in this PR.

Reply:
@bakkot I don't think we should have inconsistency in the preambles, and I don't think we should have an intermediate state where we have no specific description of the underlying operation. What do you think we should do here?

Reply:
@michaelficarra I don't see how this PR makes things worse than the current state, and also don't really understand what you'd consider better.

Reply:
Sure, it doesn't make things worse. I've opened https://github.com/tc39/ecma262/issues/1988 to track this.